bokehjs/src/lib/core/util/templating.ts

```diff
@@ -3,84 +3,144 @@ import * as Numbro from "numbro"
 
-function _format_number(num: string | number): string {
-  if (isNumber(num)) {
+export type FormatterType = "numeral" | "printf" | "datetime"
+export type FormatterSpec = CustomJSHover | FormatterType
+export type Formatters = {[key: string]: FormatterSpec} | null
+export type FormatterFunc = (value:any, format:string, special_vars: Vars) => string
+export type Index = number | ImageIndex
+export type Vars = {[key: string]: any}
+
+export const DEFAULT_FORMATTERS = {
+  "numeral"  : function(value:any, format:string, _special_vars: Vars) { return Numbro.format(value, format) },
+  "datetime" : function(value:any, format:string, _special_vars: Vars) { return tz(value, format)            },
+  "printf"   : function(value:any, format:string, _special_vars: Vars) { return sprintf(format, value)       },
+}
+
+export function basic_formatter(value:any, _format:string, _special_vars: Vars): string {
+  if (isNumber(value)) {
     const format = (() => {
       switch (false) {
-        case Math.floor(num) != num:
+        case Math.floor(value) != value:
           return "%d"
-        case !(Math.abs(num) > 0.1) || !(Math.abs(num) < 1000):
+        case !(Math.abs(value) > 0.1) || !(Math.abs(value) < 1000):
           return "%0.3f"
         default:
           return "%0.3e"
       }
     })()
 
-    return sprintf(format, num)
-  } else
-    return `${num}`  // get strings for categorical types
+    return sprintf(format, value)
+  }
+
+  else
+    return `${value}`  // get strings for categorical types
+}
+
+export function get_formatter(name: string, raw_spec: string, format: string, formatters: Formatters = null) : FormatterFunc {
+  // no format, use default built in formatter
+  if (format==null)
+    return basic_formatter
+
+  // format spec in the formatters dict, use that
+  if (formatters != null && (name in formatters || raw_spec in formatters)) {
+
+    // some day (Bokeh 2.0) we can get rid of the check for name, and just check the raw spec
+    // keep it now for compatibility but do not demonstrate it anywhere
+    let key: string
+    if (raw_spec in formatters)
+      key = raw_spec
+    else
+      key = name
+    const formatter = formatters[key]
+
+    if (isString(formatter)) {
+      if (formatter in DEFAULT_FORMATTERS)
+        return DEFAULT_FORMATTERS[formatter]
+      else
+        throw new Error(`Unknown tooltip field formatter type '${formatter}'`)
+    }
+    return function(value:any, format:string, special_vars: Vars) : string { return formatter.format(value, format, special_vars) }
+  }
+
+  // otherwise use "numeral" as default
+  return DEFAULT_FORMATTERS["numeral"]
+
+}
+
+export function get_value(name: string, data_source: ColumnarDataSource, i: Index, special_vars: Vars) {
+
+  if (name[0] == "$")
+    if (name.substring(1) in special_vars)
+      return special_vars[name.substring(1)]
+    else
+      throw new Error(`Unknown special variable '${name}'`)
```

Comment:
This needs to be either flattened or put in curly brackets.

Reply:
Addressed in d0b6c9cc5

Reply:
Actually I was speaking only about the outer block and not the inner blocks. There isn't usually value in using curly brackets with one liners (maybe besides loops; I'm not consistent with this across all classes of statements).

Reply:
Now the braces have been added only for the outermost block in ced4d9adc